File: Kap13/Python/Listing39_47/MyGraphs.py
import abc
import math

class GrafikElm (metaclass=abc.ABCMeta):

    @abc.abstractmethod
    def __init__(self, randfarbe):
        self._randfarbe = randfarbe
        self._logger = None
        print("GrafikElm erzeugt")

    def registerLogger(self, logger):
        self._logger = logger

    def getRandfarbe(self):
        return self._randfarbe

    def setRandfarbe(self, randfarbe):
        self._randfarbe = randfarbe
    
    def printInfo(self):
        print("Randfarbe: " + self._randfarbe)


class FlaechenElm (GrafikElm, metaclass=abc.ABCMeta):

    @abc.abstractmethod
    def __init__(self, randfarbe, fuellfarbe):
        super().__init__(randfarbe)
        self._fuellfarbe = fuellfarbe
        self._flaeche = 0
        print("FlaechenElm erzeugt")

    @abc.abstractmethod
    def berechneFlaeche(self):
        return
        
    def getFuellfarbe(self):
        return self._fuellfarbe

    def setFuellfarbe(self, fuellfarbe):
        self._fuellfarbe = fuellfarbe
    
    def getFlaeche(self):
        return self._flaeche

    def printInfo(self):
        super().printInfo()
        print("Fuellfarbe: " + self._fuellfarbe)
        print("Flaeche: " + self._flaeche)


class ZweiDElm (FlaechenElm, metaclass=abc.ABCMeta):

    @abc.abstractmethod
    def __init__(self, randfarbe, fuellfarbe):
        FlaechenElm.__init__(self, randfarbe, fuellfarbe)
        self._umfang = 0
        print("ZweiDElm erzeugt")

    @abc.abstractmethod
    def berechneUmfang(self):
        return    
    
    def getUmfang(self):
        if  self._logger != None :
            self._logger.doLog("Umfang wurde abgefragt.");
        return self._umfang;        

    def printInfo(self):
        super().printInfo()
        print("Umfang: " + self._umfang)
        

class Jsonifyable (metaclass=abc.ABCMeta):

    @abc.abstractmethod
    def __init__(self, randfarbe, fuellfarbe):
        return
    
    @abc.abstractmethod
    def getJsonString(self):
        return
    

class Dreieck (ZweiDElm):

    def __init__(self, randfarbe, fuellfarbe, a, b, c):
        super().__init__(randfarbe, fuellfarbe)
        print("Start Konstruktor Dreieck")
        if self.checkValues(a, b, c):
            self.__seiteA = a
            self.__seiteB = b
            self.__seiteC = c
            self.berechneFlaeche()
        print("Ende Konstruktor Dreieck")

    def checkValues(self, a, b, c) :
        if a + b <= c or b + c <= a or c + a <= b :
            return False
        return True
    
    def berechneUmfang(self):
        self._umfang = self.__seiteA + self.__seiteB + self.__seiteC

    def berechneFlaeche(self):
        self.berechneUmfang()
        s = self._umfang / 2
        self._flaeche = math.sqrt(s * (s - self.__seiteA) * (s - self.__seiteB) * (s - self.__seiteC))
    
    def getSeiteA(self):
        return self.__seiteA
    
    def getSeiteB(self):
        return self.__seiteB
    
    def getSeiteC(self):
        return self.__seiteC
    
    def setSeiteA(self, a):
        if self.checkValues(a, self.__seiteB, self.__seiteC):
            self.__seiteA = a
    
    def setSeiteB(self, b):
        if self.checkValues(self.__seiteA, b, self.__seiteC):
            self.__seiteB = b
    
    def setSeiteC(self, c):
        if self.checkValues(self.__seiteA, self.__seiteB, c):
            self.__seiteC = c

    def printInfo(self):
        super().printInfo()
        print("Seite A: " + self.__seiteA)
        print("Seite B: " + self.__seiteB)
        print("Seite C: " + self.__seiteC)
        

class Kreis (ZweiDElm):

    def __init__(self, randfarbe, fuellfarbe, radius):
        super().__init__(randfarbe, fuellfarbe)
        print("Start Konstruktor Kreis")
        self.__radius = radius
        self.berechneUmfang()
        self.berechneFlaeche()
        print("Ende Konstruktor Kreis")
    
    def berechneUmfang(self):
        self._umfang = 2 * self.__radius * math.pi

    def berechneFlaeche(self):
        self._flaeche = self.__radius * self.__radius * math.pi
    
    def getRadius(self):
        return self.__radius
    
    def setRadius(self, radius):
        self.__radius = radius
    
    def printInfo(self):
        super().printInfo()
        print("Radius: " + self.__radius)


class Rechteck (ZweiDElm, Jsonifyable):

    def __init__(self, randfarbe, fuellfarbe, lanege, breite):
        ZweiDElm.__init__(randfarbe, fuellfarbe)
        self.__laenge = lanege
        self.__breite = breite
        self.berechneUmfang()
        self.berechneFlaeche()
        print("Rechteck erzeugt")
    
    def getJsonString(self):
        jsonRep = "{\"Typ: \":\"Rechteck\"" + ",\n\"Randfarbe\":" + self._randfarbe + ",\n\"Fuellfarbe\":" + self._fuellfarbe + ",\n\"Umfang\":" + self._umfang + ",\n\"Flaeche\":" + self._flaeche + ",\n\"Laenge\":" + self.__laenge + ",\n\"Breite\":" + self.__breite + "}";
        return jsonRep
    
    def berechneUmfang(self):
        self._umfang = 2 * self.__lanege + 2 * self.__breite

    def berechneFlaeche(self):
        self._flaeche = self.__lanege * self.__breite
    
    def getLaenge(self):
        return self.__laenge
    
    def getBreite(self):
        return self.__breite
    
    def setLaenge(self, laenge):
        self.__laenge = laenge
    
    def setBreite(self, breite):
        self.__breite = breite
    
    def printInfo(self):
        super().printInfo()
        print("Laenge: " + self.__laenge)
        print("Breite: " + self.__breite)

File: Kap13/PHP/Listing39_47/MyGraphs.php
<?php
namespace mygraphs;


interface Jsonifyable {
  public function getJsonString();
}

abstract class GrafikElm {
    protected $randfarbe;
    protected $logger = null;

    public function __construct($initRandfarbe) {
        $this->randfarbe = $initRandfarbe;
        echo "\nGrafikElm erzeugt.";
    }

    public function registerLogger($logger) {
      $this->logger = $logger;
    }

    public function getRandfarbe() {
      return $this->randfarbe;
    }
    public function setRandfarbe($newRandfarbe) {
      $this->randfarbe = $newRandfarbe;
    }
    public function printInfo() {
      echo "\nRandfarbe: " .$this->randfarbe;
    }
}

abstract class FlaechenElm extends GrafikElm {
  protected $fuellfarbe;
  protected $flaeche = 0;
   
  public function __construct($initRandfarbe, $initFuellfarbe) {
    parent::__construct($initRandfarbe);
    $this->fuellfarbe = $initFuellfarbe;
    echo "\nFlaechenElm erzeugt.";
  }
 
  abstract protected function berechneFlaeche();

  public function getFuellfarbe() {
    return $this->fuellfarbe;
  }

  public function setFuellfarbe($newFuellfarbe) {
    $fuellfarbe = $newFuellfarbe;
  }

  public function getFlaeche() {
    return $this->flaeche;
  }
  public function printInfo() {
    parent::printInfo();
    echo "Fuellfarbe: " .$fthis->uellfarbe;
    echo "Flaeche: " .$this->flaeche;
  }
}

abstract class ZweiDElm extends FlaechenElm {
  protected $umfang = 0;
   
  public function __construct($initRandfarbe, $initFuellfarbe) {
    parent::__construct($initRandfarbe, $initFuellfarbe);
    echo "\nZweiDElm erzeugt.";
  }
 
  abstract protected function berechneUmfang();

  public function getUmfang() {
    if ($this->logger != null) {
      $this->logger->doLog("Umfang wurde abgefragt.");
    }
    return "\n" .$this->umfang;
  }


  public function printInfo() {
    parent::printInfo();
    echo "\nUmfang: " .$this->umfang;
  }
}	

class Kreis extends ZweiDElm {
    protected $radius;
     
    public function __construct($initRandfarbe, $initFuellfarbe, $radius) {
      parent::__construct($initRandfarbe, $initFuellfarbe);
      $this->radius = $radius;
      $this->berechneUmfang();
      $this->berechneFlaeche();
      echo "\nKreis erzeugt.";
    }
    
    protected function berechneUmfang() {
      $this->umfang = 2*$this->radius * M_PI;
    }

    protected function berechneFlaeche() {
      $this->flaeche = $this->radius * $this->radius * M_PI;
    }

    public function setRadius($radius) {
        $this->radius = $radius;
    }
    
    public function getRadius() {
      return radius;
    }

      public function printInfo() {
      parent::printInfo();
      echo "\nRadius: " .$this->radius;
    }
}

class Rechteck extends ZweiDElm implements Jsonifyable {
  private $laenge;
  private $breite;
   
  public function __construct($initRandfarbe, $initFuellfarbe, $laenge, $breite) {
    parent::__construct($initRandfarbe, $initFuellfarbe);
    $this->laenge = $laenge;
    $this->breite = $breite;
    $this->berechneUmfang();
    $this->berechneFlaeche();
    System.out.println("Rechteck erzeugt.");
  }

  public function getJsonString() {
    $jsonRep = 
      "{\"Typ: \":\"Rechteck\""
      .",\n\"Randfarbe\":" .randfarbe
      .",\n\"Fuellfarbe\":" .fuellfarbe
      .",\n\"Umfang\":" .umfang
      .",\n\"Flaeche\":" .flaeche
      .",\n\"Laenge\":" .laenge
      .",\n\"Breite\":" .breite ."}";
    return jsonRep;
  }

  protected function berechneUmfang() {
    $this->umfang = 2*$this->laenge + 2*$this->breite;
  }

  protected function berechneFlaeche() {
    $this->flaeche = $this->laenge * $this->breite;
  }

  public function getLaenge() {
    return $this->laenge;
  }
    
  public function setLaenge($laenge) {
      $this->laenge = $laenge;
  }

    public function getBreite() {
      return $this->breite;
    }
    
    public function setBreite($breite) {
      $this->breite = $breite;
    }
  
  public function printInfo() {
    parent::printInfo();
    echo "\nLaenge: " . $this->laenge;
    echo "\nBreite: " . $this->breite;
  }
}

final class Dreieck extends ZweiDElm {
  private $seiteA = 1;
  private $seiteB = 1;
  private $seiteC = 1;

    public function __construct($initRandfarbe, $initFuellfarbe, $a, $b, $c) {
      parent::__construct($initRandfarbe, $initFuellfarbe);
      echo "\nStart Konstruktor Dreieck";
      if ($this->checkValues($a, $b, $c)) {
        $this->seiteA = $a;
        $this->seiteB = $b;
        $this->seiteC = $c;
        $this->berechneFlaeche();        
      }
      echo "\nEnde Konstruktor Dreieck";
    }
    
    private function checkValues($a, $b, $c) {
      if ($a + $b <= $c || $b + $c <= $a || $c + $a <= $b) {
        return false;
      }
      return true;
    }

    protected function berechneUmfang() {
      $this->umfang = $this->seiteA + $this->seiteB + $this->seiteC;
    }

    protected function berechneFlaeche() {
      $this->berechneUmfang();
        $s = $this->umfang/2;
        $this->flaeche = sqrt($s * ($s - $this->seiteA) * ($s - $this->seiteB) * ($s - $this->seiteC));
    }

    public function getSeiteA() {
      return $this->seiteA;
    }
    
    public function setSeiteA($a) {
      if (checkValues($a, $this->seiteB, $this->seiteC)) {
        $this->seiteA = $a;
        $this->berechneFlaeche();
      }
    }
    public function getSeiteB() {
      return $this->seiteB;
    }
    
    public function setSeiteB($b) {
      if (checkValues($this->seiteA, $b, $this->seiteC)) {
        $this->seiteB = $b;
        $this->berechneFlaeche();
      }
    }
    public function getSeiteC() {
      return $this->seiteC;
    }
    
    public function setSeiteC($c) {
      if (checkValues($this->seiteA, $this->seiteB, $c)) {
        $this->seiteC = $c;
        $this->berechneFlaeche();
      }
    }
    
    public function printInfo() {
      parent::printInfo();
      echo "\nSeite A: " + $this->seiteA;
      echo "\nSeite B: " + $this->seiteB;
      echo "\nSeite C: " + $this->seiteC;
   }
  
}

?>